File: 2020/3/both.pl
#!/usr/bin/perl

use strict;
use integer;

my $fh;

die "* no input\n" if (!open($fh, '<', $ARGV[0] || 'input.txt'));

my $linenum = 0;
my @map;

while (my $line = <$fh>)
{
	$linenum++;
	if ($line =~ /^([\.#]+)\s*$/)
	{
		push @map, $1;
	}
}

my $answer = count_trees(\@map, 3, 1);

printf("Part 1 answer: %d\n", $answer);

$answer *= count_trees(\@map, 1, 1);
$answer *= count_trees(\@map, 5, 1);
$answer *= count_trees(\@map, 7, 1);
$answer *= count_trees(\@map, 1, 2);

printf("Part 2 answer: %d\n", $answer);

sub count_trees
{
	my ($map, $dx, $dy) = @_;
	my $width = length($map->[0]);
	my $height = @$map;
	my $trees = 0;
	my $x = 0;
	my $y = 0;

	while ($y < $height)
	{
		$x += $dx;
		$y += $dy;
		$trees++ if (substr($map->[$y], $x % $width, 1) eq '#');
	}

	return $trees;
}

File: 2020/3/both.py
#!/usr/bin/python

import re
import string
import sys

infile = 'input.txt'
if (len(sys.argv) > 1):
	infile = sys.argv[1]

fh = open(infile, 'r');

map_re = re.compile(r"^([\.#]+)\s*$")

map = []

for line in fh:
	m = map_re.match(line)
	if m:
		map.append(m.group(1))

def count_trees(map, dx, dy):
	width = len(map[0])
	height = len(map)
	trees = 0
	x = 0
	y = 0
	while True:
		x += dx
		y += dy
		if y >= height:
			break
		if map[y][x % width] == '#':
			trees += 1
	return trees

answer = count_trees(map, 3, 1)

print("Part 1 answer: %d" % answer)

answer *= count_trees(map, 1, 1)
answer *= count_trees(map, 5, 1)
answer *= count_trees(map, 7, 1)
answer *= count_trees(map, 1, 2)

print("Part 2 answer: %d" % answer)
